File: LWHeroOC/Classes/UIKit+Hero.m
//
//  UIKit+Hero.m
//  Hero-ObjectiveC
//
//  Created by luca.li on 2017/1/22.
//  Copyright © 2017年 luca. All rights reserved.
//

#import "UIKit+Hero.h"
#import "Hero.h"
#import <objc/runtime.h>

static NSString *parameterRegex = @"(?:\\-?\\d+(\\.?\\d+)?)|\\w+";
static NSString *modifiersRegex = @"(\\w+)(?:\\(([^\\)]*)\\))?";

struct AssociatedKeys {
    __unsafe_unretained NSString *HeroID;
    __unsafe_unretained NSString *HeroModifiers;
};

static struct AssociatedKeys assoKey = {@"ht_heroID", @"ht_heroModifers"};

@implementation UIView (Hero)

@dynamic heroID;
@dynamic heroModifiers;
@dynamic heroModifierString;

- (NSString *)heroID {
    return objc_getAssociatedObject(self, &assoKey.HeroID);
}

- (void)setHeroID:(NSString *)heroID {
    objc_setAssociatedObject(self,
                             &assoKey.HeroID,
                             heroID,
                             OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (NSArray <HeroModifier *>*)heroModifiers {
    return objc_getAssociatedObject(self, &assoKey.HeroModifiers);
}

- (void)setHeroModifiers:(NSArray <HeroModifier *>*)heroModifiers {
    objc_setAssociatedObject(self,
                             &assoKey.HeroModifiers,
                             heroModifiers,
                             OBJC_ASSOCIATION_RETAIN_NONATOMIC
                             );
}

- (NSString *)heroModifierString {
    NSAssert(0, @"Reverse lookup is not supported");
    return nil;
}

- (void)setHeroModifierString:(NSString *)heroModifierString {
    NSString *modifierString = heroModifierString;
    NSMutableArray <HeroModifier*>*modifiers = [NSMutableArray array];
    
    for (NSTextCheckingResult *r in [self matchesForRegex:modifiersRegex text:modifierString]) {
        NSMutableArray <NSString *>*parameters = [NSMutableArray array];
        if (r.numberOfRanges > 2 && [r rangeAtIndex:2].location < modifierString.length) {
            NSString *parameterString = [modifierString substringWithRange:[r rangeAtIndex:2]];
            for (NSTextCheckingResult *r in [self matchesForRegex:parameterRegex text:parameterString]) {
                [parameters addObject:[parameterString substringWithRange:r.range]];
            }
        }
        NSString *name = [modifierString substringWithRange:[r rangeAtIndex:1]];
        HeroModifier *modifier = [HeroModifier modifierFromName:name parameters:parameters];
        if (modifier) {
            [modifiers addObject:modifier];
        }
    }
    
    self.heroModifiers = modifiers;
}

- (UIView *)slowSnapshotView {
    UIGraphicsBeginImageContextWithOptions(self.bounds.size, self.isOpaque, 0);
    [self.layer renderInContext:UIGraphicsGetCurrentContext()];
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    UIImageView *imageView = [[UIImageView alloc] initWithImage:image];
    imageView.frame = self.bounds;
    UIView *snapshotView = [[UIView alloc] initWithFrame:self.bounds];
    [snapshotView addSubview:imageView];
    
    return snapshotView;
}

#pragma mark - Private

- (NSArray <NSTextCheckingResult *>*)matchesForRegex:(NSString *)regex text:(NSString *)text {
    NSError *error;
    NSRegularExpression *regularExpression = [NSRegularExpression regularExpressionWithPattern:regex options:NSRegularExpressionCaseInsensitive error:&error];
    
    if (error) {
        NSLog(@"invalid regex: \(error.localizedDescription)");
        return @[];
    } else {
        return [regularExpression matchesInString:text options:NSMatchingReportCompletion range:NSMakeRange(0, text.length)];
    }
}

@end

@implementation NSObject (Archiver)

- (id)copyWithArchiver {
    return [NSKeyedUnarchiver unarchiveObjectWithData:[NSKeyedArchiver archivedDataWithRootObject:self]];
}

@end


struct PreviousDelegates {
    __unsafe_unretained NSString *navigationDelegate;
    __unsafe_unretained NSString *tabBarDelegate;
};

static struct PreviousDelegates prevDelegates = {@"heroPreviousNavigationDelegate", @"heroPreviousTabBarDelegate"};

@implementation UIViewController (Delegate)

- (id<UINavigationControllerDelegate>)previousNavigationDelegate {
    return objc_getAssociatedObject(self, &prevDelegates.navigationDelegate);
}

- (void)setPreviousNavigationDelegate:(id<UINavigationControllerDelegate>)previousNavigationDelegate {
    objc_setAssociatedObject(self,
                             &prevDelegates.navigationDelegate,
                             previousNavigationDelegate,
                             OBJC_ASSOCIATION_RETAIN_NONATOMIC
                             );
}

- (id<UITabBarControllerDelegate>)previousTabBarDelegate {
    return objc_getAssociatedObject(self, &prevDelegates.tabBarDelegate);
}

- (void)setPreviousTabBarDelegate:(id<UITabBarControllerDelegate>)previousTabBarDelegate {
    objc_setAssociatedObject(self,
                             &prevDelegates.tabBarDelegate,
                             previousTabBarDelegate,
                             OBJC_ASSOCIATION_RETAIN_NONATOMIC
                             );
}

- (BOOL)isHeroEnabled {
    return [self.transitioningDelegate isKindOfClass:[Hero class]];
}

- (void)setIsHeroEnabled:(BOOL)isHeroEnabled {
    if (isHeroEnabled) {
        self.transitioningDelegate = [Hero shared];
        UINavigationController *navi = (UINavigationController *)self;
        if (navi && [navi isKindOfClass:[UINavigationController class]]) {
            self.previousNavigationDelegate = navi.delegate;
            navi.delegate = [Hero shared];
        }
        UITabBarController *tab = (UITabBarController *)self;
        if (tab && [tab isKindOfClass:[UITabBarController class]]) {
            self.previousTabBarDelegate = tab.delegate;
            tab.delegate = [Hero shared];
        }
    } else {
        self.transitioningDelegate = nil;
        UINavigationController *navi = (UINavigationController *)self;
        if (navi && [navi.delegate isKindOfClass:[Hero class]]) {
            navi.delegate = self.previousNavigationDelegate;
        }
        UITabBarController *tab = (UITabBarController *)self;
        if (tab && [tab.delegate isKindOfClass:[Hero class]]) {
            tab.delegate = self.previousTabBarDelegate;
        }
    }
}

- (IBAction)ht_dismiss:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)heroReplaceViewControllerWithNext:(UIViewController *)next {
    UINavigationController *navigationController = self.navigationController;
    if (navigationController) {
        NSMutableArray *vcs = [NSMutableArray array];
        vcs = [navigationController.childViewControllers mutableCopy];
        if ([vcs count] > 0) {
            [vcs removeLastObject];
            [vcs addObject:next];
        }
        
        if (navigationController.isHeroEnabled) {
            [Hero shared].forceNotInteractive = @(YES);
        }
        [navigationController setViewControllers:vcs animated:YES];
    } else if (self.view.superview) {
        UIViewController *parentVC = [self presentingViewController];
        UIView *container = self.view.superview;
        id <UIViewControllerTransitioningDelegate> oldTransitionDelegate = [next transitioningDelegate];
        next.isHeroEnabled = YES;
        [[Hero shared] transitionFrom:self to:next inView:container completion:^(BOOL finished){
            if (![oldTransitionDelegate isKindOfClass:[Hero class]]) {
                next.isHeroEnabled = NO;
                next.transitioningDelegate = oldTransitionDelegate;
            }
            
            if (finished) {
                [[UIApplication sharedApplication].keyWindow addSubview:[next view]];
                if (parentVC) {
                    [self dismissViewControllerAnimated:NO completion:^{
                        [parentVC presentViewController:next animated:NO completion:nil];
                    }];
                } else {
                    [[UIApplication sharedApplication].keyWindow setRootViewController:next];
                }
            }
        }];
    }
}
@end

@implementation UIImage (ScreenShot)

- (UIImage *)imageWithView:(UIView *)view {
    UIGraphicsBeginImageContextWithOptions(view.bounds.size, NO, 0.0);
    [view drawViewHierarchyInRect:view.bounds afterScreenUpdates:YES];
    UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return img;
}

@end

File: LWHero_swift/LWHero/Classes/UIKit+Hero.swift
//
//  UIKit+Hero.swift
//  LWHero
//
//  Created by Swift Port
//  Copyright © 2025. All rights reserved.
//

import Foundation
import UIKit
import ObjectiveC

// MARK: - UIView + Hero

private var heroIDKey: UInt8 = 0
private var heroModifiersKey: UInt8 = 1
private var heroModifierStringKey: UInt8 = 2

extension UIView {
    @IBInspectable public var heroID: String? {
        get {
            return objc_getAssociatedObject(self, &heroIDKey) as? String
        }
        set {
            objc_setAssociatedObject(self, &heroIDKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)
        }
    }

    public var heroModifiers: [HeroModifier]? {
        get {
            return objc_getAssociatedObject(self, &heroModifiersKey) as? [HeroModifier]
        }
        set {
            objc_setAssociatedObject(self, &heroModifiersKey, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }

    @IBInspectable public var heroModifierString: String? {
        get {
            return objc_getAssociatedObject(self, &heroModifierStringKey) as? String
        }
        set {
            objc_setAssociatedObject(self, &heroModifierStringKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)
            if let newValue = newValue {
                heroModifiers = parseModifierString(newValue)
            } else {
                heroModifiers = nil
            }
        }
    }

    public func slowSnapshotView() -> UIView {
        UIGraphicsBeginImageContextWithOptions(bounds.size, isOpaque, 0)
        defer { UIGraphicsEndImageContext() }

        if let context = UIGraphicsGetCurrentContext() {
            layer.render(in: context)
        }

        let image = UIGraphicsGetImageFromCurrentImageContext()
        let imageView = UIImageView(image: image)
        imageView.frame = bounds

        return imageView
    }

    private func parseModifierString(_ string: String) -> [HeroModifier] {
        var modifiers: [HeroModifier] = []
        let components = string.components(separatedBy: " ")

        var i = 0
        while i < components.count {
            let name = components[i]
            var parameters: [String] = []

            i += 1
            while i < components.count && !components[i].hasSuffix(")") {
                let param = components[i].trimmingCharacters(in: CharacterSet(charactersIn: "(),"))
                if !param.isEmpty {
                    parameters.append(param)
                }
                i += 1
            }

            if i < components.count {
                let lastParam = components[i].trimmingCharacters(in: CharacterSet(charactersIn: "(),"))
                if !lastParam.isEmpty {
                    parameters.append(lastParam)
                }
                i += 1
            }

            if let modifier = HeroModifier.modifier(fromName: name, parameters: parameters) {
                modifiers.append(modifier)
            }
        }

        return modifiers
    }
}

// MARK: - UIViewController + Hero

private var previousNavigationDelegateKey: UInt8 = 10
private var previousTabBarDelegateKey: UInt8 = 11
private var isHeroEnabledKey: UInt8 = 12

extension UIViewController {
    public var previousNavigationDelegate: UINavigationControllerDelegate? {
        get {
            return objc_getAssociatedObject(self, &previousNavigationDelegateKey) as? UINavigationControllerDelegate
        }
        set {
            objc_setAssociatedObject(self, &previousNavigationDelegateKey, newValue, .OBJC_ASSOCIATION_ASSIGN)
        }
    }

    public var previousTabBarDelegate: UITabBarControllerDelegate? {
        get {
            return objc_getAssociatedObject(self, &previousTabBarDelegateKey) as? UITabBarControllerDelegate
        }
        set {
            objc_setAssociatedObject(self, &previousTabBarDelegateKey, newValue, .OBJC_ASSOCIATION_ASSIGN)
        }
    }

    @IBInspectable public var isHeroEnabled: Bool {
        get {
            return objc_getAssociatedObject(self, &isHeroEnabledKey) as? Bool ?? false
        }
        set {
            objc_setAssociatedObject(self, &isHeroEnabledKey, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)

            if newValue {
                if let navigationController = navigationController {
                    previousNavigationDelegate = navigationController.delegate
                    navigationController.delegate = Hero.shared
                }

                if let tabBarController = tabBarController {
                    previousTabBarDelegate = tabBarController.delegate
                    tabBarController.delegate = Hero.shared
                }

                transitioningDelegate = Hero.shared
            } else {
                if let navigationController = navigationController,
                   navigationController.delegate === Hero.shared {
                    navigationController.delegate = previousNavigationDelegate
                }

                if let tabBarController = tabBarController,
                   tabBarController.delegate === Hero.shared {
                    tabBarController.delegate = previousTabBarDelegate
                }

                if transitioningDelegate === Hero.shared {
                    transitioningDelegate = nil
                }
            }
        }
    }

    @IBAction public func ht_dismiss(_ sender: Any?) {
        dismiss(animated: true, completion: nil)
    }

    public func heroReplace(with next: UIViewController) {
        if let navigationController = navigationController {
            var viewControllers = navigationController.viewControllers
            if let index = viewControllers.firstIndex(of: self) {
                viewControllers[index] = next
                navigationController.setViewControllers(viewControllers, animated: false)
            }
        }
    }
}

// MARK: - Array Helper Extension

extension Array {
    func get(at index: Int) -> Element? {
        guard index >= 0 && index < count else { return nil }
        return self[index]
    }

    func getCGFloat(at index: Int) -> CGFloat? {
        guard let element = get(at: index) else { return nil }

        if let value = element as? CGFloat {
            return value
        } else if let value = element as? Double {
            return CGFloat(value)
        } else if let value = element as? Float {
            return CGFloat(value)
        } else if let value = element as? Int {
            return CGFloat(value)
        } else if let string = element as? String {
            return Double(string).map { CGFloat($0) }
        }

        return nil
    }

    func getDouble(at index: Int) -> Double? {
        guard let element = get(at: index) else { return nil }

        if let value = element as? Double {
            return value
        } else if let value = element as? CGFloat {
            return Double(value)
        } else if let value = element as? Float {
            return Double(value)
        } else if let value = element as? Int {
            return Double(value)
        } else if let string = element as? String {
            return Double(string)
        }

        return nil
    }

    func getBool(at index: Int) -> Bool? {
        guard let element = get(at: index) else { return nil }

        if let value = element as? Bool {
            return value
        } else if let string = element as? String {
            return string.lowercased() == "true" || string == "1"
        } else if let number = element as? NSNumber {
            return number.boolValue
        }

        return nil
    }
}
